tapchannelmsg/custom_channel_data.go

```diff
@@ -0,0 +1,302 @@
+package tapchannelmsg
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/taproot-assets/rfqmsg"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/tlv"
+	"google.golang.org/protobuf/proto"
+)
+
+// ReadOpenChannel reads the content of an OpenChannel struct from a reader.
+func ReadOpenChannel(r io.Reader) (*OpenChannel, error) {
+	openChanData, err := wire.ReadVarBytes(
+		r, 0, tlv.MaxRecordSize, "chan data",
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read open chan data: %w", err)
+	}
+
+	var openChannelRecord OpenChannel
+	err = openChannelRecord.Decode(bytes.NewReader(openChanData))
+	if err != nil {
+		return nil, fmt.Errorf("error decoding custom channel data: %w",
+			err)
+	}
+
+	return &openChannelRecord, nil
+}
+
+// ReadCommitment reads the content of a Commitment struct from a reader.
+func ReadCommitment(r io.Reader) (*Commitment, error) {
+	localCommitData, err := wire.ReadVarBytes(
+		r, 0, tlv.MaxRecordSize, "commit data",
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read open chan data: %w", err)
+	}
+
+	var localCommit Commitment
+	err = localCommit.Decode(bytes.NewReader(localCommitData))
+	if err != nil {
+		return nil, fmt.Errorf("error decoding custom commit data: %w",
+			err)
+	}
+
+	return &localCommit, nil
+}
+
+// ChannelCustomData represents the data that is returned in the
+// CustomChannelData field of a lnrpc.Channel object.
+type ChannelCustomData struct {
+	OpenChan    OpenChannel
+	LocalCommit Commitment
+}
+
+// AsJson returns the JSON representation of the channel custom data.
```

Comment:
why do we bother touching JSON? i'm missing something, this is not user-exposed data right?

Reply:
It's to render the custom TLV encoded data as JSON on the CLI and RPC/REST interfaces. So it takes the TLV data and renders it as JSON on the RPC output, so on the CLI for example we just have to turn the raw bytes into a string and format the nested JSON.